File: _experimental/_c++/numbers/src/implementations/numbers_iterative.cpp
#include "../interfaces/numbers_iterative.h"

int NumbersIterative::fibonacci(int n)
{
    return fibonacci_iter(n, 0, 1);
}

int NumbersIterative::fibonacci_iter(int n, int acc2, int acc1)
{
    if (n <= 1)
    {
        return n;
    }
    else if(n == 2)
    {
        return acc1 +acc2;
    }
    else
    {
        return fibonacci_iter(n - 1, acc1, acc1 + acc2);
    }
}

int NumbersIterative::factorial(int n)
{
    return factorial_iter(n, 1);
}

int NumbersIterative::factorial_iter(int n, int accumulator)
{
    if(n <= 1)
    {
        return accumulator;
    }
    return factorial_iter(n - 1, n * accumulator);
}

int NumbersIterative::sum_numbers(int n)
{
    return sum_numbers_iter(n, 0);
}

int NumbersIterative::sum_numbers_iter(int n, int accumulator)
{
    if(n <= 0){
        return accumulator;
    }
    return sum_numbers_iter(n - 1, n + accumulator);
}

File: _experimental/_erlang/numbers/src/numbers_iterative.erl
-module(numbers_iterative).
-export([fibonacci/1, factorial/1, sum_numbers/1]).

fibonacci(N) ->
    fibonacci_iter(N, 0, 1).

fibonacci_iter(N, _, _) when N =< 1 ->
    N;
fibonacci_iter(N, Acc2, Acc1) when N =:= 2 ->
    Acc1 + Acc2;
fibonacci_iter(N, Acc2, Acc1) ->
    fibonacci_iter(N - 1, Acc1, Acc1 + Acc2).

factorial(N) ->
    factorial_iter(N, 1).

factorial_iter(N, Accumulator) when N =< 1 ->
    Accumulator;
factorial_iter(N, Accumulator) ->
    factorial_iter(N - 1, N * Accumulator).

sum_numbers(N) ->
    sum_numbers_iter(N, 0).

sum_numbers_iter(N, Accumulator) when N =< 0 ->
    Accumulator;
sum_numbers_iter(N, Accumulator) ->
    sum_numbers_iter(N - 1, N + Accumulator).